arrays/student_validator_using_pointers.c

```diff
@@ -0,0 +1,267 @@
+#include <stdio.h>
+#include <time.h>
+
+typedef struct
+{
+    int day;
+    int month;
+    int year;
+} Date;
+
+typedef struct 
+{
+    Date date_of_birth;
+    char sex;
+} PersonDate;
+
+int validate_input(PersonDate* card)
+{
+    return 
+    !(
+        &(*card).date_of_birth.year < 1901 || 
+        &(*card).date_of_birth.day > 31 || 
+        &(*card).date_of_birth.month > 12 || 
+        &(*card).date_of_birth.day < 1 || 
+        &(*card).date_of_birth.month < 1
+    ) || (
+        &(*card).sex != 'M' && 
+        &(*card).sex != 'F'
+    );
+}
+
+int calculate_age(int birthYear) 
+{
+    time_t t = time(NULL);
+    struct tm tm = *localtime(&t);
+    int this_year = tm.tm_year + 1900;
+    return this_year - birthYear;
+}
+
+
+
+int check_card(PersonDate card) 
+{
+    int birth = card.date_of_birth.year;
+    int age = calculate_age(card.date_of_birth.year);
+    char sex = card.sex;
+    return validate_input(&card) && age >= 18 && age <= 55;
+}
+
+void feed_back(PersonDate card)
+{
+    int birth = card.date_of_birth.year;
+    int age = calculate_age(card.date_of_birth.year);
+    if (check_card(card))
+    {
+        printf("Sorry, not for you %d", age);
+    } 
+    else 
+    {
+        printf("Nice... Age is: %d", age);
+    }
+}
+
+int can_engage_in_sexual_activity(PersonDate* card1, PersonDate* card2)
+{
+    int age_difference = &(*card1).date_of_birth.year - &(*card2).date_of_birth.year;
+    return (age_difference < 30 || age_difference > -30) && &(*card1).sex != &(*card2).sex;
+}
+
+char to_uppercase(char letter) 
+{
+    if (letter == 'f')
+    {
+        return 'F';
+    } 
+    else if (letter == 'm')
+    {
+        return 'M';
+    }
+    else {
+        return letter;
+    } 
+}
+
+int main()
+{
+    int birthDay;
+    int birthMonth;
+    int birthYear;
+    char sex;
+
+    int studentBirthDay;
+    int studentBirthMonth;
+    int studentBirthYear;
+    char studentSex;
+
+    int teacherBirthDay;
+    int teacherBirthMonth;
+    int teacherBirthYear;
+    char teacherSex;
+
+    int hamaliyBirthDay;
+    int hamaliyBirthMonth;
+    int hamaliyBirthYear;
+    char hamaliySex;
```

Comment:
Use structs instead of creating this tone of redundant variables. You can then use `scanf` on the needed fields of the struct objects